### ISSUE TITLE
Question about the benchmarks

### ISSUE THREAD

**Issue by tbusser** [jonschlinkert/pad-left#8]
I was looking at the code for `pad-left` and `left-pad` and was wondering about how fair the benchmarks are. 

Both `pad-left` and `left-pad` use the same sort of algorithm for creating the padding yet in the benchmarks `pad-left` is about twice as fast. The reason for this difference in speed is the `repeat-string` package. As long as the padding char remains the same it doesn't recreate the string, it simply returns a cached version.

It is a very nice optimisation which can make quite the difference but it also skews the benchmark results.  The `left-pad` implementation is actually padding the string on each iteration where `pad-left` already has part of the answer ready after the first iteration.

When benchmarking the speed of both packages, would it not be more fair if it was done without the caching of `repeat-string`? Or perhaps make sure no two iterations use the same padding character.

I feel that now the benchmarks are only running a single case which heavily favours the implementation of `pad-left`. By adding cases which would render the caching by `repeat-string` moot you would get a better representation of how fast both algorithms are.

**Comment by jonschlinkert** [comment]
You know what, I'm seriously losing hope in this community. I created this lib two years ago, then they just updated their code to be similar, now you're on THIS repo asking me why I did something wrong? 

How about asking left-pad why they didn't just use this lib instead of re-creating the same code? Nothing else needs to be said. I shouldn't have to defend something I did before the code in left-pad existed.

> Both pad-left and left-pad use the same sort of algorithm for creating the padding yet in the benchmarks pad-left is about twice as fast.

I published this two years ago using that algorithm in [repeat-string](https://github.com/jonschlinkert/repeat-string). `left-pad` is slowly updating their code to be more like this one. Should I have anticipated that they would copy the code and made different decisions based on that? 

> It is a very nice optimisation which can make quite the difference but it also skews the benchmark results.

It improves benchmark results because it makes the code more performant. It's repeating a string - not randomizing one. The point is to emulate what might happen in reality.  If you're suggesting that it's unlikely for someone to repeat a string "so much" or to repeat the same string on different function calls, that's not a debate I want to engage in.  In any case, it's a totally valid and super useful optimization. 

Still, would I remove a legit optimization to help another library? Especially one that's copying off this one? How did you get the nerve to even ask these questions? 

> would it not be more fair if it was done without the caching of repeat-string? ... By adding cases which would render the caching by repeat-string moot you would get a better representation of how fast both algorithms are.

wouldn't it be more fair for me to just focus on writing code instead of having to defend something I created two years ago against something that's copying my lib? 

Regarding the code. Did you actually thing about what you were writing before you typed out those words? The core algorithms are essentially the same!??! So 1) how would either lib be faster if the code was identical (if I removed the caching), and 2) again, why not just use this lib?!

**Comment by tbusser** [comment]
I'm sorry to hear my message made you feel so defensive, that was most certainly not my intention. It was also not at all my intention to suggest you should alter the code of your package, I only meant to suggest altering the benchmarks. I would also like to note that I was not aware of the history between `pad-left` and `left-pad` and that it was such a delicate subject. I only learned of `left-pad` through the readme of your project.

My question/suggestion was just out of curiosity. It started with me wondering what would be the fastest way to pad a string. I made a couple of implementations and started looking into a good way to compare them, this led me to benchmark.js. Based on the benchmarks I ran on my implementations the one that used Duff's device was the quickest at about 115k ops/sec on my machine. 

Then I opened Google, searched for "pad left" and found your package at the top of the search results. I installed the package, added a new benchmark and was blown away when your package did 35M ops/sec on my machine. Curious how you achieved such a vastly superior performance I started to look at the code. The caching of the result turned out the be biggest reason why. Once I implemented this in my implementation with Duff's device my padding method achieved almost the same ops/sec as yours.

On your Github page for the `pad-left` package it states it is the fastest implementation and the benchmarks with `left-pad` are featured prominently in the readme. This made me curious about their implementation. Looking at their code I noticed the way of creating the padding string also used  `if (num & 1)` and `num >>= 1`.This made me wonder if both implementations would score about the same if caching is taken out of the equation. What I wound up doing is taking the caching out of `pad-left` and benchmarked it against `left-pad` and both wound up scoring about 6M ops/sec.

The fact that `pad-left` and `left-pad` scored roughly the same when caching is taken out of the equation made me think about the value of the benchmarks. Even my crappy implementation scored almost the same ops/sec with the caching mechanism as your package but I would never think of  my implementation is being as good as yours.

Like I said, I am sorry if my message offended you. I love the package you've made, the optimization with the caching is awesome and like I wrote in the original message, it makes quite the difference in performance. I only meant to exchange ideas about the meaning and relevance of the benchmarks. 

I might be pushing it but after looking at the code I do have two questions about `repeat-string`.
1. Given the impact caching has on the performance, have you ever considered caching something like the last 5 strings to repeat? Would this add something useful in your opinion or do you feel this is something which will not help much in real world usage?
2. Inside the `while` loop there is a check for `!num` and it breaks out of the `while` loop. I was wondering if this was added for performance reasons? It doesn't seem to be strictly needed as the `while` loop already checks for `num > 0`. Furthermore, is there is specific reason you check for `!num` inside the `while` loop instead of `num === 0`?

**Comment by jonschlinkert** [comment]
I'm sorry I responded so badly. I tend to get passionate about projects I care about, and I was already on edge since you're not the first person to create an issue of a similar nature related to left-pad.

> have you ever considered caching something like the last 5 strings to repeat

It's worth a shot, but my gut tells me that it would probably have a negative impact on performance. The reason is that when caching strings we keep everything on the stack and avoid creating any references in memory. That said, v8 is getting so optimized that it might not matter as much anymore. I guess try adding objects anywhere in the function and see what happens. It will either be noticeably slower or nothing lol.

> Once I implemented this in my implementation with Duff's device my padding method achieved almost the same ops/sec as yours.

Really? That's surprising, as I recall none of the other algorithms in the benchmarks were even close. But I don't recall if I was using caching at the time. anyway that's a good pattern. I like the combo of while/switch statements

> Inside the while loop there is a check for !num and it breaks out of the while loop. I was wondering if this was added for performance reasons?

Yes, since we're a) a bit-shifting `num` on the previous line, it might be `0` at that point, b) building up the string exponentially, and c) slicing off excess when the string is returned, the longer the string is the more benefit there is to breaking as early as possible. Put another way, if we don't break there, the string will effectively double in size on the next line after it has already reached it's max length, and it's cheaper to check `num` than it is for `.substr` to search a longer string. This is why I check twice - both times are just before the string is doubled.

the caveat with caching is that it needs to be implemented correctly so that that cache is cleared when the pattern changes and, if necessary, provide a way to clear the cache from outside the closure. otherwise you're just creating memory leaks. 

> Is there is specific reason you check for !num inside the while loop instead of num === 0?

No it would probably be better to use `num === 0`, since strict equality is always faster. It wouldn't make much difference, but if I wrote the code today I'd use `num === 0`

**Comment by tbusser** [comment]
Thank you for your long and detailed answer! I really appreciate you taking the time to answer my questions and I found the answers to be really useful. 

> Really? That's surprising, as I recall none of the other algorithms in the benchmarks were even close. 

That's funny because it was the part that didn't surprise me at all. The benchmark was now only incurring the performance penalty of my slower repeat string algorithm once instead of each run.

I am still trying to wrap my head around how the `repeat-string` algorithm works. I understand the purpose of each line, I know what each line does and still I am unable to explain it in plain English. The part I am struggling with the most is where you add the value of `str` to `res`. Why do you have to do this each time the least significant bit is set? I guess my math skills are letting me down here. Is there any reading material on the subject you can recommend?

I really enjoyed reading your code and I am in still in awe of the performance you managed to achieve. Thanks for making it open source and giving others the opportunity to learn something.

**Comment by jonschlinkert** [comment]
> That's funny because it was the part that didn't surprise me at all.

one thing that really impacts the benchmarks is the optimizations that have happened in v8 over the past couple of years. When I created this lib, `while` or `for` loops were substantially faster (orders of magnitude) than native methods like `.forEach`. but today the gap has closed significantly, to the point where native methods and loops are often equal.
